keps/sig-node/xxx-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,1372 @@
+documentation such as release notes or a development roadmap. It should be
+possible to collect this information before implementation begins, in order to
+avoid requiring implementors to split their attention between writing release
+notes and implementing the feature itself. KEP editors and SIG Docs
+should help to ensure that the tone and content of the `Summary` section is
+useful for a wide audience.
+
+A good summary is probably at least a paragraph in length.
+
+Both in this section and below, follow the guidelines of the [documentation
+style guide]. In particular, wrap lines to a reasonable length, to make it
+easier for reviewers to cite specific portions, and to minimize diff churn on
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+Dynamic resource allocation introduces a new API for requesting
+arbitrary resources with parameters that can be different depending on
+what kind of resource gets requested. Resource allocations can be
+ephemeral (specified in a Pod spec, allocated and destroyed together
+with the Pod) and persistent (lifecycle managed separately from a Pod
+and used for multiple different Pods).
+
+Resources are managed by plugins that communicate with central
+Kubernetes components, in particular the kube-scheduler, by updating
+objects stored in the kube-apiserver. Communication with kubelet is
+handled through local Unix domain sockets and the plugin registration
+mechanism, using a new plugin type and a new gRPC interface. Towards
+the container runtime, kubelet and those plugins then must use the
+[Container Device Interface
+(CDI)](https://github.com/container-orchestrated-devices/container-device-interface)
+to expose the resources.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+Originally, Kubernetes and its scheduler only tracked CPU and RAM as
+resources for containers. Later, support for storage and discrete,
+countable per-node extended resources was added. The device plugin
+interface then made such local resources available to containers. But
+for many newer devices, this approach and the Kubernetes API for
+requesting these custom resources are too limited:
+
+- *Device initialization*: When starting a workload I’d like to have
+  the device reconfigured or reprogrammed by the orchestration. For
+  security reasons workloads should not be able to reconfigure devices
+  directly.
+
+  *Limitation*: Currently, it’s impossible to specify the desired
+  device properties that are required for reconfiguring devices.
+
```

Comment:
When looking at the current device plugin API, being able to add hooks (e.g. preContainerStart) in addition to devices, mounts, and envvars would allow us to run dynamic initialisation and checks in the container. One example in the case of the current NVIDIA implementation is that we run ldconfig to update the ld.cache in the container and have the injected libraries visible to the application container. These hooks could also include checks for compatible library versions as defined by an image vendor.

Reply:
This leads back to the overall question: should we pursue this KEP or focus instead on incremental improvements of the existing device plugin API?

@elezar: any thoughts on this? Do you have use cases now or in the future that simply can't be handled unless we do something completely new, like this KEP?

Reply:
In general, incremental improvements have a higher likelihood of being accepted. It doesn't mean you can't eventually end up in the same final state that you want, it just means that the path to get there might not be as "big-bang" as originally proposed. I've seen lots of KEPs that propose large changes just fizzle out and die because its too much too fast (even if the underlying ideas were actually pretty good). Funnily enough, it will often move faster if you propose a bunch of incremental changes because it will be more obvious to the community that each incremental change is useful in its own right.